options_linux.cmake

```diff
@@ -53,10 +53,12 @@ if (DESKTOP_APP_USE_PACKAGED)
 else()
     find_library(ATOMIC_LIBRARY libatomic.a)
 endif()
-
-if (ATOMIC_LIBRARY)
-    target_link_libraries(common_options
-    INTERFACE
-        ${ATOMIC_LIBRARY}
-    )
+if (NOT ATOMIC_LIBRARY)
+    # Last resort when find_library() gives up.
+    set(ATOMIC_LIBRARY atomic)
```

Comment:
this will add an unwanted shared dependency to official binary and break it on a lot of installations

Reply:
I don't get why this can break the binary that you name so pretentiously. Can you please give some examples? And overall, how were these `if`s supposed to work?

Reply:
> I don't get why this can break the binary that you name so pretentiously.

https://github.com/telegramdesktop/tdesktop/issues/8299

Reply:
Well, you could use `--as-needed` liner feature to eliminate redundant runtime dependecies.

Reply:
> how were these `if`s supposed to work?

Exactly how they were before your changes. If we're building in packaged mode, we're checking for shared libatomic pretense, if it's present, then we're linking to it. If it's not, we assume we on an old system without it or we're building on an architecture where it's not really needed and the builder doesn't want to link to it. For non-packaged mode it's the same except of libatomic should be static.

Reply:
cmake successfully finds the library on my machine. You either have a misconfigured cmake build or a buggy cmake version.

Reply:
Could you please tell me what path exactly does it find? What value is assigned to the ATOMIC_LIBRARY variable?

Reply:
Last time I checked building on my host a couple of months ago and I don't want to waste time to check this, so I can't

Reply:
Very good. Then it worked for you earlier. And now it doesn't work in April. Otherwise I can't explain why CMake should find the library in a private GCC directory, <code>/usr/lib/<b>gcc</b>/bla-bla-blah/...</code>

Reply:
Anyway, find_library is supposed to find a library and if it can't, it is some bug in cmake or in your distro

Reply:
At least you could check it with `find /usr -name libatomic.so -or -name libatomic.a` or a similar command with mlocate(1).

Reply:
I checked the presence in your distro in the time of discussion in the tdesktop contributions group
https://packages.debian.org/search?searchon=contents&keywords=libatomic.so&mode=path&suite=stable&arch=any

PS: my distro doesn't have /usr :P

Reply:
> Otherwise I can't explain why CMake should find the library in a private GCC directory, `/usr/lib/gcc/bla-bla-blah/...`

it is not in private directory in Arch for instance
https://archlinux.org/packages/core/x86_64/gcc-libs/

Reply:
it's also not in private directory in my distro: `/nix/store/gzcjh911w4s81j0gzbz90rwyyfxcb23x-gcc-10.2.0-lib/lib/libatomic.so
`

Reply:
> PS: my distro doesn't have /usr :P

So you use somehow patched CMake or it was specially configured for your OS. You can't expect the same from others.

As for original issue, can we give a hint to CMake so that it would be able to discover correct path to the libatomic?

Reply:
> So you use somehow patched CMake or it was specially configured for your OS. You can't expect the same from others.

So you entirely skipped the arch link

Reply:
I didn't. Arch Linux put private libraries to a public directory, didn't it?

Reply:
Yeah. So it is a problem of your distro.

Reply:
Alas, the actually bug is in this options_linux.cmake file. It access to the private library without necessary check or search hint.

Reply:
Anyway, your PR breaks the official binary

Reply:
That's because your _official_ binary is built only for x86_64 which does not require libatomic for wide types.

Reply:
Your PR still breaks it

Reply:
Then look at my previous PR which really checked need for libatomic. https://github.com/telegramdesktop/tdesktop/pull/7080

Reply:
I don't see any problem in the current checking code. If it doesn't work for you then just point cmake where it should search or patch the check in your distro patches.

Reply:
Just pass `-latomic` flag to linker. That is all build scripts have to do. And compiler would be able to find the library.

Reply:
This is not acceptable since the logic should depend on whether the library is present or not or official binary will be broken.

Reply:
> the logic should depend on whether the library is present or not

That's wrong. Scripts should check `std::atomic` applicability or use the `--as-needed` option as I already mentioned above.

Reply:
That's just your opinion. I've seen that other projects link without the check and without `--as-needed`. Please stop this and just specify path to the library in your cmake arguments.

Reply:
So is this script incomplete and buggy and so you suggest me such crappy workaround?

Reply:
The script can't do magic and know where the private directory in your distro. At least I don't know a way.

Reply:
Okay. I think such pointing should be integrated into the script. So one could invoke cmake without unusual arguments.

Reply:
> I think such pointing should be integrated into the script.

What do you mean? Please don't reinvent CMAKE_LIBRARY_PATH.

Reply:
It is very weird restriction.